var/spack/repos/builtin/packages/libtiff/package.py

```diff
@@ -103,6 +103,12 @@ class Libtiff(CMakePackage, AutotoolsPackage):
     # https://gitlab.com/libtiff/libtiff/-/merge_requests/243
     conflicts("platform=darwin", when="@4.3.0")
 
+    # cmake 3.19 and newer turn on pic automatically when shared is enabled
+    # for older cmake versions enabling shared without pic results in
+    # relocation R_X86_64_32 against `.rodata.str1.1' can not
+    # be used when making a shared object; recompile with -fPIC
+    conflicts("~pic", when="+shared ^cmake@:3.18")
```

Comment:
Wondering if we shouldn't do this for all versions of cmake. It sounds like `libtiff+shared+pic` and `libtiff+shared~pic` install the exact same thing in CMake 3.19+, is that correct?

Reply:
Yes, it seems that `cmake@3.19:` use other default flags. The flags for `libtiff+shared+pic` and `libtiff+shared~pic` are exactly the same:

`libtiff+shared~pic ^cmake 3.26`
```
cd $spack-stage/spack-stage-libtiff-4.5.0-6gmn7zckboygte7gfkjdjnjcjnotyxzs/spack-build-6gmn7zc/libtiff && $spack/lib/spack/env/gcc/gcc -D_FILE_OFFSET_BITS=64 -Dtiff_EXPORTS -I$spack-stage/spack-stage-libtiff-4.5.0-6gmn7zckboygte7gfkjdjnjcjnotyxzs/spack-src/libtiff -I$spack-stage/spack-stage-libtiff-4.5.0-6gmn7zckboygte7gfkjdjnjcjnotyxzs/spack-build-6gmn7zc/libtiff -isystem $spack/opt/spack/linux-fedora37-x86_64/gcc-12.3.1/zlib-1.2.13-uex44uq55uqoq34xmkh3kibt6mvi3fhz/include -isystem $spack/opt/spack/linux-fedora37-x86_64/gcc-12.3.1/libjpeg-turbo-2.1.5-ioyru3li43kh76552cifgjvdb5o42z7t/include -Wall -Winline -Wformat-security -Wpointer-arith -Wdisabled-optimization -Wno-unknown-pragmas -fstrict-aliasing -O3 -DNDEBUG -std=gnu99 -fPIC -MD -MT libtiff/CMakeFiles/tiff.dir/tif_close.c.o -MF CMakeFiles/tiff.dir/tif_close.c.o.d -o CMakeFiles/tiff.dir/tif_close.c.o -c $spack-stage/spack-stage-libtiff-4.5.0-6gmn7zckboygte7gfkjdjnjcjnotyxzs/spack-src/libtiff/tif_close.c
```

`libtiff+shared+pic ^cmake 3.26`
```
cd $spack-stage/spack-stage-libtiff-4.5.0-rkbepdug3e2dzri7zsz47qwvsjoxq2kp/spack-build-rkbepdu/libtiff && $spack/lib/spack/env/gcc/gcc -D_FILE_OFFSET_BITS=64 -Dtiff_EXPORTS -I$spack-stage/spack-stage-libtiff-4.5.0-rkbepdug3e2dzri7zsz47qwvsjoxq2kp/spack-src/libtiff -I$spack-stage/spack-stage-libtiff-4.5.0-rkbepdug3e2dzri7zsz47qwvsjoxq2kp/spack-build-rkbepdu/libtiff -isystem $spack/opt/spack/linux-fedora37-x86_64/gcc-12.3.1/zlib-1.2.13-uex44uq55uqoq34xmkh3kibt6mvi3fhz/include -isystem $spack/opt/spack/linux-fedora37-x86_64/gcc-12.3.1/libjpeg-turbo-2.1.5-ioyru3li43kh76552cifgjvdb5o42z7t/include -Wall -Winline -Wformat-security -Wpointer-arith -Wdisabled-optimization -Wno-unknown-pragmas -fstrict-aliasing -O3 -DNDEBUG -std=gnu99 -fPIC -MD -MT libtiff/CMakeFiles/tiff.dir/tif_close.c.o -MF CMakeFiles/tiff.dir/tif_close.c.o.d -o CMakeFiles/tiff.dir/tif_close.c.o -c $spack-stage/spack-stage-libtiff-4.5.0-rkbepdug3e2dzri7zsz47qwvsjoxq2kp/spack-src/libtiff/tif_close.c
```

I didn't add the conflict for all `cmake` versions to have it documented that something changed in cmake. If you want we can open it up.

Reply:
@alalazo @svenevs any idea what might have changed in CMake such that:
```
-DBUILD_SHARED_LIBS=ON -DCMAKE_POSITION_INDEPENDENT_CODE=ON
```
would work in CMake 3.19+ but not in older CMake? Does `+shared~pic` even make sense in general for packages? I know nothing of the compiled world.

Reply:
Any update here? Or shall we merge it as is?

Reply:
> Does +shared~pic even make sense in general for packages?

I don't think so. Usually I recommend people to allow the `pic` variant only `when=~shared`.

Reply:
That solution sounds cleaner to me too.